horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,173 @@
+                                reader_pool_type=self.reader_pool_type,
+                                workers_count=self.train_reader_worker_count,
+                                shard_count=hvd.size(),
+                                hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+                                schema_fields=self.schema_fields,
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                shuffle_rows=self.shuffle,
+                                shuffle_row_groups=self.shuffle,
+                                seed=self.random_seed,
+                                **self.reader_factory_kwargs)
+        self.val_reader = self.reader_factory(self.val_dir,
+                                num_epochs=1,
+                                cur_shard=hvd.rank(),
+                                reader_pool_type=self.reader_pool_type,
+                                workers_count=self.val_reader_worker_count,
+                                shard_count=hvd.size(),
+                                hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+                                schema_fields=self.schema_fields,
+                                transform_spec=self.transform_spec,
+                                storage_options=self.storage_options,
+                                shuffle_rows=False,
+                                shuffle_row_groups=False,
+                                **self.reader_factory_kwargs) if self.has_val else self.empty_batch_reader()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.stop()
+        if self.train_reader:
+            self.train_reader.stop()
+        super().__exit__(type, value, traceback)
+
+    @contextlib.contextmanager
+    def empty_batch_reader(self):
+        yield None
+
+    def train_data(self):
+        return self.make_dataset(self.train_reader, self.train_batch_size,
+                                 self.is_batch_reader, shuffle=self.shuffle, cache=self.inmemory_cache_all)
+
+    def val_data(self):
+        return self.make_dataset(self.val_reader, self.val_batch_size,
+                                 self.is_batch_reader, shuffle=False, cache=self.inmemory_cache_all) if self.val_reader else None
+
+
+class NVTabularDataModule(DataModule):
+    """NVTabular-based DataModule for KerasEstimator for GPU-accelerated data loading of tabular datasets.
+
+    Note: requires `label_cols`, `categorical_cols`, and `continuous_cols` to be explicitly provided."""
+    def __init__(self, label_cols=[], categorical_cols=[], continuous_cols=[], **kwargs):
+        super().__init__(**kwargs)
+        self.label_cols = label_cols
+        self.categorical_cols = categorical_cols
+        self.continuous_cols = continuous_cols
+
+    @staticmethod
+    def seed_fn():
+        """
```

Comment:
Question to NVTabular specific usage: maybe we should use `random_seed` in constructor?
In that case we don't need `hvd.broadcast` here right?
User can specify `random_seed` for reproducible results.

Reply:
This `seed_fn` is invoked on each worker to seed and re-seed the shuffle operation (consistently across workers), so probably should leave as is.

Reply:
SG. Leave as it.